source/client.c

```diff
@@ -501,8 +513,112 @@ static void s_outstanding_request_destroy(void *item) {
+            AWS_CONTAINER_OF(node, struct aws_mqtt_outstanding_request, list_node);
+        /* Fire the callback and clean up the memory, as the connection get destoried. */
+        if (!request->completed) {
+            request->on_complete(
+                request->connection, request->packet_id, AWS_ERROR_MQTT_CONNECTION_DESTROYED, request->on_complete_ud);
+        }
+        aws_memory_pool_release(&connection->synced_data.requests_pool, request);
+    }
+
+    AWS_LOGF_DEBUG(AWS_LS_MQTT_CLIENT, "id=%p: Destroying connection", (void *)connection);
+
+    /* If the reconnect_task isn't freed, free it */
+    if (connection->reconnect_task) {
+        aws_mem_release(connection->reconnect_task->allocator, connection->reconnect_task);
+    }
+    aws_string_destroy(connection->host_name);
 
+    /* Clear the credentials */
+    if (connection->username) {
+        aws_string_destroy_secure(connection->username);
+    }
+    if (connection->password) {
+        aws_string_destroy_secure(connection->password);
+    }
+
+    /* Clean up the will */
+    aws_byte_buf_clean_up(&connection->will.topic);
+    aws_byte_buf_clean_up(&connection->will.payload);
+
+    /* Clear the client_id */
+    aws_byte_buf_clean_up(&connection->client_id);
+
+    /* Free all of the active subscriptions */
+    aws_mqtt_topic_tree_clean_up(&connection->thread_data.subscriptions);
+
+    aws_hash_table_clean_up(&connection->synced_data.outstanding_requests_table);
+    aws_memory_pool_clean_up(&connection->synced_data.requests_pool);
+
+    aws_mutex_clean_up(&connection->synced_data.lock);
+
+    aws_tls_connection_options_clean_up(&connection->tls_options);
+
+    /* Clean up the websocket proxy options */
+    if (connection->websocket.proxy) {
+        aws_tls_connection_options_clean_up(&connection->websocket.proxy->tls_options);
+
+        aws_mem_release(connection->allocator, connection->websocket.proxy);
+        connection->websocket.proxy = NULL;
+        connection->websocket.proxy_options = NULL;
+    }
+
+    aws_mqtt_client_release(connection->client);
+
+    /* Frees all allocated memory */
+    aws_mem_release(connection->allocator, connection);
+}
+
+static void s_on_final_disconnect(struct aws_mqtt_client_connection *connection, void *userdata) {
+    (void)userdata;
+
+    s_mqtt_client_connection_destroy_final(connection);
+}
+
+static void s_mqtt_client_connection_start_destroy(struct aws_mqtt_client_connection *connection) {
+    bool call_destroy_final = false;
+
+    { /* BEGIN CRITICAL SECTION */
+        mqtt_connection_lock_synced_data(connection);
+        if (connection->synced_data.state != AWS_MQTT_CLIENT_STATE_DISCONNECTED) {
+
+            /*
+             * We don't call the on_disconnect callback until we've transitioned to the DISCONNECTED state.  So it's
+             * safe to change it now while we hold the lock since we know we're not DISCONNECTED yet.
+             */
```

Comment:
I know I said it should be safe to do this before... 
However, I just realized a crazy corner case. The problem is we are not protecting the callback by a lock, since we don't want to invoke any callback with holding a lock.  So, it's very unlikely, but possible that if the state is DISCONNECTED, `aws_mqtt_client_connection_connect` called and switch the state to CONNECTING before the `on_disconnect` invoked. We will crash...
We can probably document it like "don't call connect if you call disconnect and the callback has not been invoked"?

Reply:
I don't think this is a case we need to worry about.  We're only executing this logic if every single reference to the connection has been released.  Nothing stop the user from releasing and then using the pointer but that's a usage error on them, not us.